### PR TITLE
[flutter_tool] Clean up usage events and custom dimensions

### PR DESCRIPTION
**Pull request by zanderso**
## Description

- Changes GA custom dimensions to an enum.
- Begins using classes to enforce proper usage of custom dimensions.
- Sends target platform, etc. with all hot reload events, not just exceptions.

This PR also changes the code under reporting/ from a number of separate libraries into one library split into separate 'part' files under reporting/reporting.dart. This will allow making use of annotations like `@visibleForOverriding` (although it doesn't work yet https://github.com/dart-lang/sdk/issues/29804).

## Related Issues

https://github.com/flutter/flutter/issues/36566

## Tests

I added the following tests:

Added test that successful hot reload event sends target platform etc.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by zanderso** [comment]
ptal

File: packages/flutter_tools/lib/src/reporting/events.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+  final String category;
+  final String parameter;
+
+  void send() {
+    flutterUsage.sendEvent(category, parameter);
+  }
+}
+
+class HotEvent extends UsageEvent {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/reporting/events.dart]: I think the contextual comments you left in the runner might also make sense as doc comments on this class too

**Review by jonahwilliams — APPROVED** [review]
LGTM

Thanks for this change, it really makes the analytics api nicer to work with

File: packages/flutter_tools/lib/src/reporting/events.dart (hunk 1 above)
Comment (zanderso) [reply packages/flutter_tools/lib/src/reporting/events.dart]: Included those comments here, and added comments in various other places as well.

**Comment by sethladd** [comment]
Thanks very much!

**Comment by Hixie** [comment]
We should not use `part of`:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-library-and-part-of

Is there some other way to solve this?

**Comment by zanderso** [comment]
@Hixie I'd prefer not dumping code all together in the same file. In particular, I don't want to move the code from events.dart into usage.dart. The alternative to using 'part of' to achieve this is to import/re-export libraries. Unfortunately, doing that breaks the visibility annotations understood by the analyzer. That is, when in the same library, the code in events.dart can call Usage.sendEvent() even though it is annotated @visibleForTesting, which is what I want. I want the code in events.dart to be the only code that can call Usage.sendEvent(). But when different files implies different libraries, the code in events.dart can no longer call Usage.sendEvent() (or I can no longer give it a visibility annotation, and everyone can call it willy-nilly).

**Pull request merged**